src/utils/cache.h

```diff
@@ -31,116 +31,132 @@
-      if (key == iter->key) return true;
+    size_t idx = key % hash_.size();
+    while (true) {
+      if (hash_[idx].metadata == 0) break;
+      if (hash_[idx].metadata == 1 && hash_[idx].key == key) {
+        return true;
+      }
+      ++idx;
+      if (idx >= hash_.size()) idx -= hash_.size();
     }
     return false;
   }
 
   // Looks up and pins the element by key. Returns nullptr if not found.
-  // If found, brings the element to the head of the queue (makes it last to
-  // evict); furthermore, a call to Unpin must be made for each such element.
-  // Use of LruCacheLock is recommended to automate this pin management.
-  V* LookupAndPin(K key) {
+  // If found, a call to Unpin must be made for each such element.
+  // Use of HashedKeyCacheLock is recommended to automate this pin management.
+  V* LookupAndPin(uint64_t key) {
     if (capacity_.load(std::memory_order_relaxed) == 0) return nullptr;
 
     SpinMutex::Lock lock(mutex_);
 
-    auto hash = hasher_(key) % hash_.size();
-    for (Item* iter = hash_[hash]; iter; iter = iter->next_in_hash) {
-      if (key == iter->key) {
-        // BringToFront(iter);
-        ++iter->pins;
-        return iter->value.get();
+    size_t idx = key % hash_.size();
+    while (true) {
+      if (hash_[idx].metadata == 0) break;
+      if (hash_[idx].metadata == 1 && hash_[idx].key == key) {
+        ++hash_[idx].pins;
+        return hash_[idx].value.get();
       }
+      ++idx;
+      if (idx >= hash_.size()) idx -= hash_.size();
     }
     return nullptr;
   }
 
-  // Unpins the element given key and value. Use of LruCacheLock is recommended
-  // to automate this pin management.
-  void Unpin(K key, V* value) {
+  // Unpins the element given key and value. Use of HashedKeyCacheLock is
+  // recommended to automate this pin management.
+  void Unpin(uint64_t key, V* value) {
     SpinMutex::Lock lock(mutex_);
 
     // Checking evicted list first.
-    Item** cur = &evicted_head_;
-    for (Item* el = evicted_head_; el; el = el->next_in_hash) {
-      if (key == el->key && value == el->value.get()) {
-        if (--el->pins == 0) {
-          *cur = el->next_in_hash;
+    for (auto it = evicted_.begin(); it != evicted_.end(); ++it) {
+      auto& entry = *it;
+      if (key == entry.key && value == entry.value.get()) {
+        if (--entry.pins == 0) {
           --allocated_;
-          delete el;
+          evicted_.erase(it);
+          return;
+        } else {
+          return;
         }
-        return;
       }
-      cur = &el->next_in_hash;
     }
-
-    // Now lookup in active list.
-    auto hash = hasher_(key) % hash_.size();
-    for (Item* iter = hash_[hash]; iter; iter = iter->next_in_hash) {
-      if (key == iter->key && value == iter->value.get()) {
-        assert(iter->pins > 0);
-        --iter->pins;
+    // Now the main list.
+    size_t idx = key % hash_.size();
+    while (true) {
+      if (hash_[idx].metadata == 0) break;
```

Comment:
`assert(hash_[idx].metadata == 1)` ?

Reply:
Once its a bool - that'll be implicitly covered by the assert(false) outside the loop.